src/librustc/mir/interpret/value.rs

```diff
@@ -237,13 +237,18 @@ impl<'tcx, Tag> Scalar<Tag> {
     }
 
     #[inline]
-    pub fn from_uint(i: impl Into<u128>, size: Size) -> Self {
+    pub fn try_from_uint(i: impl Into<u128>, size: Size) -> Option<Self> {
         let i = i.into();
-        assert_eq!(
-            truncate(i, size), i,
-            "Unsigned value {:#x} does not fit in {} bits", i, size.bits()
-        );
-        Scalar::Raw { data: i, size: size.bytes() as u8 }
+        if truncate(i, size) == i {
+            Some(Scalar::Raw { data: i, size: size.bytes() as u8 })
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn from_uint(i: impl Into<u128>, size: Size) -> Self {
+        Self::try_from_uint(i, size).unwrap()
```

Comment:
Hm, might be nice to maintain the original panic message here?

Reply:
sure, i'll `.expect()` it then.

Reply:
`.expect` will compute the format string even if there is no error.

Isn't there a lazy variant of that?

Reply:
Ah, there is: use `unwrap_or_else(|| panic!("..."))`.

Reply:
I had to change the `ImmTy::from_(u)int` functions to use `Scalar::from_(u)int` directly instead of using `ImmTy::try_from_(u)int` to avoid duplicating the error message.

Reply:
Yes that makes sense.